patroni/postgresql.py

```diff
@@ -165,49 +113,85 @@ def data_directory_empty(self):
         os.path.exists(self.trigger_file) and os.unlink(self.trigger_file)
 
-    def write_pgpass(self, record):
-        with open(self.pgpass, 'w') as f:
-            os.fchmod(f.fileno(), 0o600)
-            f.write('{host}:{port}:*:{user}:{password}\n'.format(**record))
-
-        env = os.environ.copy()
-        env['PGPASSFILE'] = self.pgpass
-        return env
-
     def sync_from_leader(self, leader):
         r = parseurl(leader.conn_url)
 
-        env = self.write_pgpass(r)
+        pgpass = 'pgpass'
+        with open(pgpass, 'w') as f:
+            os.fchmod(f.fileno(), 0o600)
+            f.write('{host}:{port}:*:{user}:{password}\n'.format(**r))
+
+        env = os.environ.copy()
+        env['PGPASSFILE'] = pgpass
         return self.create_replica(r, env) == 0
 
     @staticmethod
     def build_connstring(conn):
-        return "host={host} port={port} user={user}".format(**conn)
+        mconn = ""
+        for param, val in conn.iteritems():
+            mconn = mconn + "{0}={1} ".format(param, val)
+            
+        return mconn
 
     def create_replica(self, master_connection, env):
-        self.set_state('building replica from {host}:{port}'.format(**master_connection))
+        # create the replica according to the replica_method
+        # defined by the user.  this is a list, so we need to
+        # loop through all methods the user supplies
         connstring = self.build_connstring(master_connection)
-        cmd = self.config['restore']
-        try:
-            ret = subprocess.call(shlex.split(cmd) + [self.scope, "replica", self.data_dir, connstring], env=env)
-            self.delete_trigger_file()
-        except:
-            logger.exception('Error when creating replica')
-            ret = 1
-        if ret != 0:
-            self.set_state('failed to build replica from {host}:{port}'.format(**master_connection))
-        return ret
+        env = os.environ.copy()
+        env['PGPASSFILE'] = 'pgpass'
+        # get list of replica methods from config
+        replica_list = self.config.get('create_replica_method', 'basebackup')
+        replica_methods = [rm.strip() for rm in replica_list.split(',')]
+        # go through them in priority order
+        for replica_method in replica_methods:
+            # if the method is basebackup, then use the built-in
+            if replica_method == "basebackup":
+                ret = self.basebackup(connstring, env)
+                if ret == 0:
+                    # if basebackup succeeds, exit with success
+                    return 0
+            else:
+                # user-defined method; check for configuration
+                # not required, actually
+                if replica_method in self.config:
+                    # look to see if the user has supplied a full command path
+                    # if not, use the method name as the command
+                    if "command" in self.config[replica_method]:
```

Comment:
It looks like you are trying to implement the case when replica_method is set to some 'program_name', but there is no special configuration section for it...
In this case you will get cmd undefined....